src/main/java/net/minecraftforge/fmp/block/BlockMultipartContainer.java

```diff
@@ -0,0 +1,591 @@
+package net.minecraftforge.fmp.block;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
+
+import net.minecraft.block.Block;
+import net.minecraft.block.BlockContainer;
+import net.minecraft.block.ITileEntityProvider;
+import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
+import net.minecraft.block.properties.PropertyBool;
+import net.minecraft.block.state.BlockStateContainer;
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.particle.ParticleManager;
+import net.minecraft.client.renderer.block.model.IBakedModel;
+import net.minecraft.client.renderer.block.model.ModelResourceLocation;
+import net.minecraft.client.renderer.texture.TextureAtlasSprite;
+import net.minecraft.entity.Entity;
+import net.minecraft.entity.EntityLivingBase;
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.item.ItemStack;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.BlockRenderLayer;
+import net.minecraft.util.EnumBlockRenderType;
+import net.minecraft.util.EnumFacing;
+import net.minecraft.util.EnumHand;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.math.AxisAlignedBB;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.RayTraceResult;
+import net.minecraft.util.math.Vec3d;
+import net.minecraft.world.IBlockAccess;
+import net.minecraft.world.World;
+import net.minecraft.world.WorldServer;
+import net.minecraftforge.common.property.IExtendedBlockState;
+import net.minecraftforge.common.property.IUnlistedProperty;
+import net.minecraftforge.common.util.RayTraceUtils;
+import net.minecraftforge.fml.client.FMLClientHandler;
+import net.minecraftforge.fml.relauncher.Side;
+import net.minecraftforge.fml.relauncher.SideOnly;
+import net.minecraftforge.fmp.client.multipart.AdvancedParticleManager;
+import net.minecraftforge.fmp.client.multipart.MultipartStateMapper;
+import net.minecraftforge.fmp.multipart.IMultipart;
+import net.minecraftforge.fmp.multipart.Multipart;
+import net.minecraftforge.fmp.multipart.MultipartRegistry;
+import net.minecraftforge.fmp.multipart.PartState;
+
+/**
+ * A final class that extends {@link BlockContainer} and represents a block which can contain any kind of multipart.
+ * <br/>
+ * You do NOT need to extend this class for your multiparts to work. I repeat, you do NOT. You need to either extend
```

Comment:
Also, why does this extend BlockContainer? That class is old/vanilla crap and shouldn't be used ever really.

The providing of tile entities is already handled natively by Block.hasTileEntity and Block.getTileEntity, (which are arguably superior in that they take states instead of meta)

Same for below